src/rmc-compiler/src/main.rs

```diff
@@ -0,0 +1,223 @@
+        .version_short("?")
+        .arg(
+            Arg::with_name("rmc-lib")
+                .long("--rmc-lib")
+                .value_name("FOLDER_PATH")
+                .help("Sets the path to locate the rmc library.")
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("goto-c")
+                .long("--goto-c")
+                .help("Enables compilation to goto-c intermediate representation."),
+        )
+        .arg(
+            Arg::with_name("symbol-table-passes")
+                .long("--symbol-table-passes")
+                .value_name("PASS")
+                .help("Transformations to perform to the symbol table after it has been generated.")
+                .takes_value(true)
+                .use_delimiter(true)
+                .multiple(true),
+        )
+        .arg(
+            Arg::with_name("restrict-vtable-fn-ptrs")
+                .long("--restrict-vtable-fn-ptrs")
+                .help("Restrict the targets of virtual table function pointer calls."),
+        )
+        .arg(
+            Arg::with_name("sysroot")
+                .long("--sysroot")
+                .help("Override the system root.")
+                .long_help(
+                    "The \"sysroot\" is the location where RMC will look for the Rust \
+                distribution.",
+                ),
+        )
+        .arg(
+            // TODO: Move this to a cargo wrapper. This should return rmc version.
+            Arg::with_name("rustc-version")
+                .short("V")
+                .long("--version")
+                .help("Gets underlying rustc version."),
+        )
+        .arg(
+            Arg::with_name("rustc-options")
+                .help("Arguments to be passed down to rustc.")
+                .multiple(true)
+                .takes_value(true),
+        )
+}
+
+/// Main function. Configure arguments and run the compiler.
+fn main() -> Result<(), &'static str> {
+    let matches = parser().get_matches();
+
+    // Initialize the logger.
+    init_env_logger("RMC_LOG");
+
+    // Generate rustc args.
+    let rustc_args = generate_rustc_args(&matches);
+
+    // Configure queries.
+    let mut queries = QueryDb::default();
+    queries
+        .set_symbol_table_passes(matches.values_of_lossy("symbol-table-passes").unwrap_or(vec![]));
+    queries.set_emit_vtable_restrictions(matches.is_present("restrict-vtable-fn-ptrs"));
+
+    // Configure and run compiler.
+    let mut callbacks = RmcCallbacks {};
+    install_ice_hook();
+    let mut compiler = RunCompiler::new(&rustc_args, &mut callbacks);
+    if matches.is_present("goto-c") {
+        compiler.set_make_codegen_backend(Some(Box::new(move |_cfg| {
+            rustc_codegen_rmc::GotocCodegenBackend::new(&Rc::new(queries))
+        })));
+    }
+    compiler.run().or(Err("Failed to compile crate."))
+}
+
+/// Empty struct since we don't support any callbacks yet.
+struct RmcCallbacks {}
```

Comment:
How will this be used once we do?

Reply:
Callbacks can be used to analyze each compilation step and it can also be used to stop compilation in earlier stages. For example, if we want to traverse the AST after parsing, we can implement a callback.